File: kharkivpy#0/code/barber.erl
-module(barber).
-export([run/2, barber/1, clients/1, simulator/2]).

-define(CUT_DUTAION, 20).

%% XXX: links, monitors etc are not used to focus on messaging!!!
%% XXX: use records instead of long tuples

run(RoomSize, Duration) ->
    % create barber
    BPid = spawn(?MODULE, barber, [?CUT_DUTAION]),
    % run simulartor with barber PID
    SPid = spawn(?MODULE, simulator, [BPid, RoomSize]),
    % spawn clients generator
    CSPid = spawn(?MODULE, clients, [SPid]),
    % create timer to close simulator after duration time
    timer:send_after(Duration, SPid, {close, CSPid}).

barber(Duration) ->
    receive
        {client, Simulator} -> 
            timer:sleep(Duration),
            Simulator ! {barber, done},
            barber(Duration)
    end.

clients(Simulator) ->
    Rnd = 7 + random:uniform(28), % XXX: you can play with constants
    receive
        close -> ok
    after Rnd ->
        Simulator ! {sim, client},
        clients(Simulator)
    end.

simulator({BPid, Status}, {RoomSize, Client, Total}) ->
    case serve_simulation(self(), {BPid, Status}, {RoomSize, Client, Total}) of 
        {S, C, T} -> simulator({BPid, S}, {RoomSize, C, T});
        {close, ClientsGenerator} ->
            ClientsGenerator ! close,
            io:format("Closed! Served clients (~p)~n", [Total]),
            ok
    end;
simulator(Barber, RoomSize) ->
    simulator({Barber, free}, {RoomSize, 0, 0}).

serve_simulation(Me, {BPid, Status}, {RoomSize, Client, Total}) ->
    receive
        {sim, client} when Status == free ->
            io:format("Client goes to barber~n"),
            BPid ! {client, Me},
            {busy, 0, Total};
        {sim, client} when Status == busy, RoomSize > Client ->
            io:format("Client goes to room, new size (~p)~n", [Client+1]),
            {busy, Client+1, Total};
        {sim, client} ->
            io:format("No free space for client, room size (~p)~n", [Client]),
            {busy, Client, Total};
        {barber, done} when Client > 0 -> 
            io:format("Take client from room, new size (~p)~n", [Client-1]),
            BPid ! {client, Me},
            {busy, Client-1, Total+1};
        {barber, done} -> 
            io:format("Barber finished, idle~n", []),
            {free, 0, Total+1};
        {close, ClientsGenerator} -> {close, ClientsGenerator}
    end.

File: kharkivpy#0/code/barber.go
package main

import (
    "fmt"
    "time"
    "math/rand"
)

const (
    CUTTING_TIME = 20
    BARBERS_AMOUNT   = 1
    HALL_SITS_AMOUNT = 3
)

type Barber struct {
    val int
}

type Client struct {
    val int
}

func main() {
    clients := make(chan *Client)
    go clientProducer(clients)
    go BarberShop(clients)
    time.Sleep(2 * time.Second)
}

func clientProducer(clients chan *Client) {
    for {
        time.Sleep(time.Duration(rand.Intn(28) + 7) * time.Millisecond)
        clients <- &Client{}
    }
}

func cutHear(barber *Barber, client *Client, finished chan *Barber) {
    // Cutting hear 
    time.Sleep(CUTTING_TIME * time.Millisecond)
    finished <- barber
}

func BarberShop(clients <-chan *Client) {
    freeBarbers := []*Barber{}
    waitingClient := []*Client{}
    syncBarberChan := make(chan *Barber)

    //creating barbers
    for i := 0; i < BARBERS_AMOUNT; i++ {
        freeBarbers = append(freeBarbers, &Barber{})
    }

    for {
        select {
        case client := <-clients:
            if len(freeBarbers) == 0 {
                if len(waitingClient) < HALL_SITS_AMOUNT {
                    // client is waiting in the hall
                    waitingClient = append(waitingClient, client)
                    fmt.Printf("Client is waiting in hall (%v)\n", len(waitingClient))
                } else {
                    // hall is full - bye-bye client
                    fmt.Println("No free space for client")
                }
            } else {
                barber := freeBarbers[0]
                freeBarbers = freeBarbers[1:]
                fmt.Println("Client goes to barber")
                go cutHear(barber, client, syncBarberChan)
            }
        // barber finish work
        case barber := <-syncBarberChan:
            if len(waitingClient) > 0 {
                // get client from hall
                client := waitingClient[0]
                waitingClient = waitingClient[1:]
                fmt.Printf("Take client from room (%v)\n", len(waitingClient))
                go cutHear(barber, client, syncBarberChan)
            } else {
                // barber is going to sleep
                fmt.Println("Barber idle")
                freeBarbers = append(freeBarbers, barber)
            }
        }
    }

}